pull-kubevirt-unit-test-s390x: Sometimes fails by detecting a race condition

# What happened

<!-- insert test name -->
Flaky test detected: [pull-kubevirt-unit-test-s390x](https://prow.ci.kubevirt.io/?job=pull-kubevirt-unit-test-s390x)

The unit test fails in `pkg/virt-handler/device-manager` with:
```
08:34:04: WARNING: DATA RACE
08:34:04: Write at 0x00000417f6f0 by goroutine 228:
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.init.func1.1()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/device_controller_test.go:79 +0x6eb
08:34:04:   github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func3()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/node.go:475 +0x4d
08:34:04:   github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:894 +0x16b
08:34:04: 
08:34:04: Previous read at 0x00000417f6f0 by goroutine 224:
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.initHandler()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/common.go:200 +0x49
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.discoverPermittedHostPCIDevices()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/pci_device.go:259 +0x36
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.(*DeviceController).updatePermittedHostDevicePlugins()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/device_controller.go:214 +0xbe7
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.(*DeviceController).refreshPermittedDevices()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/device_controller.go:336 +0x1d1
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.(*DeviceController).refreshPermittedDevices-fm()
08:34:04:       <autogenerated>:1 +0x7b
08:34:04: 
08:34:04: Goroutine 228 (running) created at:
08:34:04:   github.com/onsi/ginkgo/v2/internal.(*Suite).runNode()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:881 +0x14cf
08:34:04:   github.com/onsi/ginkgo/v2/internal.(*group).attemptSpec()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/group.go:199 +0x1135
08:34:04:   github.com/onsi/ginkgo/v2/internal.(*group).run()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/group.go:349 +0x13a7
08:34:04:   github.com/onsi/ginkgo/v2/internal.(*Suite).runSpecs()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:489 +0x1151
08:34:04:   github.com/onsi/ginkgo/v2/internal.(*Suite).Run()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/internal/suite.go:130 +0x677
08:34:04:   github.com/onsi/ginkgo/v2.RunSpecs()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/github.com/onsi/ginkgo/v2/core_dsl.go:300 +0xf05
08:34:04:   kubevirt.io/client-go/testutils.KubeVirtTestSuiteSetup()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/vendor/kubevirt.io/client-go/testutils/setup.go:60 +0x645
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager_test.TestDeviceManager()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/device_manager_suite_test.go:10 +0x41
08:34:04:   testing.tRunner()
08:34:04:       /root/.gimme/versions/go1.23.4.linux.s390x/src/testing/testing.go:1690 +0x261
08:34:04:   testing.(*T).Run.gowrap1()
08:34:04:       /root/.gimme/versions/go1.23.4.linux.s390x/src/testing/testing.go:1743 +0x8d
08:34:04: 
08:34:04: Goroutine 224 (finished) created at:
08:34:04:   kubevirt.io/kubevirt/pkg/virt-config.(*ClusterConfig).SetConfigModifiedCallback()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-config/configuration.go:291 +0x2f9
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.(*DeviceController).Run()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/device_controller.go:389 +0x2e9
08:34:04:   kubevirt.io/kubevirt/pkg/virt-handler/device-manager.init.func1.4.4.gowrap1()
08:34:04:       /home/prow/go/src/github.com/kubevirt/kubevirt/pkg/virt-handler/device-manager/device_controller_test.go:183 +0x8d
08:34:04: ==================
08:34:04: ••
08:34:04: 
08:34:04: Ran 31 of 34 Specs in 5.876 seconds
08:34:04: SUCCESS! -- 31 Passed | 0 Failed | 0 Pending | 3 Skipped
08:34:04:     testing.go:1399: race detected during execution of test
08:34:04: --- FAIL: TestDeviceManager (5.88s)
08:34:04: FAIL
08:34:04: FAIL	kubevirt.io/kubevirt/pkg/virt-handler/device-manager	6.081s
```

In more detail, what is happening is that the global variable `Handler` is simultaneously written to and read during a test. Looking at the code, this is caused by the unit-test and not a race condition in the kubevirt code.
It is written to during the test setup (`BeforeEach`) and read at in the normal code `initHandler()`.

This is potentially caused because tests are finishing faster than the `DeviceController` can start, causing it to still be in startup when the next testcase gets prepared.

<!-- labels for flaky tests -->
/kind flake
/priority critical-urgent

<!-- sig assignment
     all tests contain a sig identifier, please assign the corresponding SIG to the issue, i.e.
     for a test name containing [sig-compute] or [sig-operator]
-->
/sig compute

<!-- note: the flakefinder url needs to be a stable one, i.e. instead of the moving latest report use any with a date instead -->
[1]: {flakefinder_url}

## Additional context
Add any other context about the problem here.

# Flake Action Plan

**As the assignee**, **thoroughly review the issue** and put the resulting report as comment into this issue.

## Then decide on one of the following actions:

### **The flake is a bug**

* Add label commenting `/triage accepted`
* **Create a PR to fix the bug**
* Reference this issue in the PR
* Keep this issue open until the testcase does not fail anymore

### **The flake is non-critical or an issue that is hard to fix**
* **Quarantine the test** by creating a pull request assigning [QUARANTINE] tag to test name and [Quarantine decorator](https://github.com/kubevirt/kubevirt/blob/f85a8117b8a90fd913ec0719faae9506866d1525/tests/decorators/decorators.go#L7) to the test
* Reference this issue on the PR

### There was an infrastructure issue

An infra issue is anything "below" the testcase

* Add label commenting `/triage infra-issue`
* Close the issue, adding a comment with details about the infrastructure issue.

After the flake has been fixed, document the learning from it inside the fix PR

--- a/pkg/virt-handler/device-manager/common.go
+++ b/pkg/virt-handler/device-manager/common.go
@@ -55,7 +55,7 @@ type DeviceHandler interface {
 
 type DeviceUtilsHandler struct{}
 
-var Handler DeviceHandler
+var handler DeviceHandler = &DeviceUtilsHandler{}
 
 // getDeviceIOMMUGroup gets devices iommu_group
 // e.g. /sys/bus/pci/devices/0000\:65\:00.0/iommu_group -> ../../../../../kernel/iommu_groups/45
@@ -194,12 +194,6 @@ func (h *DeviceUtilsHandler) ReadMDEVAvailableInstances(mdevType string, parentI
 	}
 
 	return i, nil
-}
-
-func initHandler() {
-	if Handler == nil {
-		Handler = &DeviceUtilsHandler{}
-	}
 }
 
 func waitForGRPCServer(socketPath string, timeout time.Duration) error {

--- a/pkg/virt-handler/device-manager/device_controller.go
+++ b/pkg/virt-handler/device-manager/device_controller.go
@@ -256,7 +256,6 @@ func removeSelectorSpaces(selectorName string) string {
 	typeNameStr := strings.Replace(string(selectorName), " ", "_", -1)
 	typeNameStr = strings.TrimSpace(typeNameStr)
 	return typeNameStr
-
 }
 
 func (c *DeviceController) splitPermittedDevices(devices []Device) (map[string]Device, map[string]struct{}) {

--- a/pkg/virt-handler/device-manager/device_controller_test.go
+++ b/pkg/virt-handler/device-manager/device_controller_test.go
@@ -76,7 +76,7 @@ var _ = Describe("Device Controller", func() {
 		ctrl = gomock.NewController(GinkgoT())
 		mockPCI = NewMockDeviceHandler(ctrl)
 		mockPCI.EXPECT().GetDevicePCIID(gomock.Any(), gomock.Any()).Return("1234:5678", nil).AnyTimes()
-		Handler = mockPCI
+		handler = mockPCI
 
 		fakeConfigMap, _, _ = testutils.NewFakeClusterConfigUsingKVConfig(&v1.KubeVirtConfiguration{
 			PermittedHostDevices: &v1.PermittedHostDevices{
@@ -109,6 +109,7 @@ var _ = Describe("Device Controller", func() {
 		defer os.RemoveAll(workDir)
 		// Ensure the deviceController is stopped after each test to avoid leaking resources
 		stop <- struct{}{}
+		handler = &DeviceUtilsHandler{}
 	})
 
 	Context("Basic Tests", func() {

--- a/pkg/virt-handler/device-manager/mediated_device.go
+++ b/pkg/virt-handler/device-manager/mediated_device.go
@@ -108,8 +108,6 @@ func NewMediatedDevicePlugin(mdevs []*MDEV, resourceName string) *MediatedDevice
 	serverSock := SocketPath(mdevTypeName)
 	iommuToMDEVMap := make(map[string]string)
 
-	initHandler()
-
 	devs := constructDPIdevicesFromMdev(mdevs, iommuToMDEVMap)
 
 	dpi := &MediatedDevicePlugin{
@@ -200,8 +198,6 @@ func (dpi *MediatedDevicePlugin) Allocate(_ context.Context, r *pluginapi.Alloca
 }
 
 func discoverPermittedHostMediatedDevices(supportedMdevsMap map[string]string) map[string][]*MDEV {
-	initHandler()
-
 	mdevsMap := make(map[string][]*MDEV)
 	files, err := os.ReadDir(mdevBasePath)
 	for _, info := range files {
@@ -219,15 +215,15 @@ func discoverPermittedHostMediatedDevices(supportedMdevsMap map[string]string) m
 				typeName: mdevTypeName,
 				UUID:     info.Name(),
 			}
-			parentPCIAddr, err := Handler.GetMdevParentPCIAddr(info.Name())
+			parentPCIAddr, err := handler.GetMdevParentPCIAddr(info.Name())
 			if err != nil {
 				log.DefaultLogger().Reason(err).Errorf("failed parent PCI address for mdev: %s", info.Name())
 				continue
 			}
 			mdev.parentPciAddress = parentPCIAddr
 
-			mdev.numaNode = Handler.GetDeviceNumaNode(pciBasePath, parentPCIAddr)
-			iommuGroup, err := Handler.GetDeviceIOMMUGroup(mdevBasePath, info.Name())
+			mdev.numaNode = handler.GetDeviceNumaNode(pciBasePath, parentPCIAddr)
+			iommuGroup, err := handler.GetDeviceIOMMUGroup(mdevBasePath, info.Name())
 			if err != nil {
 				log.DefaultLogger().Reason(err).Errorf("failed to get iommu group of mdev: %s", info.Name())
 				continue

--- a/pkg/virt-handler/device-manager/mediated_device_test.go
+++ b/pkg/virt-handler/device-manager/mediated_device_test.go
@@ -105,7 +105,7 @@ var _ = Describe("Mediated Device", func() {
 			By("mocking PCI and MDEV functions to simulate an mdev an its parent PCI device")
 			ctrl = gomock.NewController(GinkgoT())
 			mockPCI = NewMockDeviceHandler(ctrl)
-			Handler = mockPCI
+			handler = mockPCI
 			// Force pre-defined returned values and ensure the function only get called exacly once each on 0000:00:00.0
 			mockPCI.EXPECT().GetMdevParentPCIAddr(fakeMdevUUID).Return(fakeAddress, nil).Times(1)
 			mockPCI.EXPECT().GetDeviceIOMMUGroup(mdevBasePath, fakeMdevUUID).Return(fakeIommuGroup, nil).Times(1)

--- a/pkg/virt-handler/device-manager/mediated_devices_types.go
+++ b/pkg/virt-handler/device-manager/mediated_devices_types.go
@@ -41,7 +41,6 @@ type MDEVTypesManager struct {
 }
 
 func NewMDEVTypesManager() *MDEVTypesManager {
-	initHandler()
 	return &MDEVTypesManager{
 		availableMdevTypesMap: make(map[string][]string),
 	}
@@ -228,14 +227,14 @@ func (m *MDEVTypesManager) configureDesiredMDEVTypes() {
 }
 
 func createMdevTypes(mdevType string, parentID string) error {
-	instances, err := Handler.ReadMDEVAvailableInstances(mdevType, parentID)
+	instances, err := handler.ReadMDEVAvailableInstances(mdevType, parentID)
 	if err != nil {
 		log.Log.Reason(err).Errorf("failed to create mdevs of type %s, failed to obtain number of instances", mdevType)
 		return err
 	}
 	// create mdevs for all available instances
 	for i := 0; i < instances; i++ {
-		err := Handler.CreateMDEVType(mdevType, parentID)
+		err := handler.CreateMDEVType(mdevType, parentID)
 		if err != nil {
 			log.Log.Reason(err).Errorf("failed to create mdevs of type %s", mdevType)
 			return err
@@ -281,7 +280,7 @@ func removeUndesiredMDEVs(desiredTypesMap map[string]struct{}) {
 	}
 	for _, file := range files {
 		if shouldRemoveMDEV(file.Name(), desiredTypesMap) {
-			err = Handler.RemoveMDEVType(file.Name())
+			err = handler.RemoveMDEVType(file.Name())
 			log.Log.Reason(err).Warningf("failed to remove mdev type: %s", file.Name())
 		}
 	}

--- a/pkg/virt-handler/device-manager/mediated_devices_types_test.go
+++ b/pkg/virt-handler/device-manager/mediated_devices_types_test.go
@@ -136,7 +136,7 @@ var _ = Describe("Mediated Devices Types configuration", func() {
 		ctrl = gomock.NewController(GinkgoT())
 		clientTest = fake.NewSimpleClientset()
 		mockMDEV = NewMockDeviceHandler(ctrl)
-		Handler = mockMDEV
+		handler = mockMDEV
 		configuredMdevTypesOnCards = make(map[string]map[string]struct{})
 
 		mockMDEV.EXPECT().CreateMDEVType(gomock.Any(), gomock.Any()).DoAndReturn(func(mdevType string, parentID string) error {

--- a/pkg/virt-handler/device-manager/pci_device.go
+++ b/pkg/virt-handler/device-manager/pci_device.go
@@ -108,8 +108,6 @@ func NewPCIDevicePlugin(pciDevices []*PCIDevice, resourceName string) *PCIDevice
 	serverSock := SocketPath(strings.Replace(resourceName, "/", "-", -1))
 	iommuToPCIMap := make(map[string]string)
 
-	initHandler()
-
 	devs := constructDPIdevices(pciDevices, iommuToPCIMap)
 
 	dpi := &PCIDevicePlugin{
@@ -256,21 +254,19 @@ func (dpi *PCIDevicePlugin) healthCheck() error {
 }
 
 func discoverPermittedHostPCIDevices(supportedPCIDeviceMap map[string]string) map[string][]*PCIDevice {
-	initHandler()
-
 	pciDevicesMap := make(map[string][]*PCIDevice)
 	err := filepath.Walk(pciBasePath, func(path string, info os.FileInfo, err error) error {
 		if info.IsDir() {
 			return nil
 		}
-		pciID, err := Handler.GetDevicePCIID(pciBasePath, info.Name())
+		pciID, err := handler.GetDevicePCIID(pciBasePath, info.Name())
 		if err != nil {
 			log.DefaultLogger().Reason(err).Errorf("failed get vendor:device ID for device: %s", info.Name())
 			return nil
 		}
 		if resourceName, supported := supportedPCIDeviceMap[pciID]; supported {
 			// check device driver
-			driver, err := Handler.GetDeviceDriver(pciBasePath, info.Name())
+			driver, err := handler.GetDeviceDriver(pciBasePath, info.Name())
 			if err != nil || driver != "vfio-pci" {
 				return nil
 			}
@@ -279,13 +275,13 @@ func discoverPermittedHostPCIDevices(supportedPCIDeviceMap map[string]string) ma
 				pciID:      pciID,
 				pciAddress: info.Name(),
 			}
-			iommuGroup, err := Handler.GetDeviceIOMMUGroup(pciBasePath, info.Name())
+			iommuGroup, err := handler.GetDeviceIOMMUGroup(pciBasePath, info.Name())
 			if err != nil {
 				return nil
 			}
 			pcidev.iommuGroup = iommuGroup
 			pcidev.driver = driver
-			pcidev.numaNode = Handler.GetDeviceNumaNode(pciBasePath, info.Name())
+			pcidev.numaNode = handler.GetDeviceNumaNode(pciBasePath, info.Name())
 			pciDevicesMap[resourceName] = append(pciDevicesMap[resourceName], pcidev)
 		}
 		return nil

--- a/pkg/virt-handler/device-manager/pci_device_test.go
+++ b/pkg/virt-handler/device-manager/pci_device_test.go
@@ -48,7 +48,7 @@ var _ = Describe("PCI Device", func() {
 		By("mocking PCI functions to simulate a vfio-pci device at " + fakeAddress)
 		ctrl = gomock.NewController(GinkgoT())
 		mockPCI = NewMockDeviceHandler(ctrl)
-		Handler = mockPCI
+		handler = mockPCI
 		// Force pre-defined returned values and ensure the function only get called exacly once each on 0000:00:00.0
 		mockPCI.EXPECT().GetDeviceIOMMUGroup(pciBasePath, fakeAddress).Return(fakeIommuGroup, nil).Times(1)
 		mockPCI.EXPECT().GetDeviceDriver(pciBasePath, fakeAddress).Return(fakeDriver, nil).Times(1)

